### ISSUE TITLE
@chrisvfritz Code Review

### ISSUE THREAD

**Issue by chrisvfritz** [egillespie/pegger-ruby-rails#1]
@egillespie Took a look. Hope this is helpful!

### General Ruby style stuff and random notes

- For variable names, `snake_case` is prefered in Ruby rather than `camelCase`
- In your models, `id` is usually preferred over something like `pegId`, as it's implied that the `id` is of the `Peg` when it's part of a `Peg` instance. 
- Semicolons at end of lines not necessary - only when you want multiple statements on the same line (e.g. `a = 1; b = 2` or `def; some_method_name; end`)
- I would delete the unused files in `app/views`. I think only `static/home.html.erb` and `layouts/application.html.erb` are being used.
- The last line of a Ruby method is implicitly returned, so most people choose not to include a `return` there. This one is a little more subjective though.
- Apart from `Integer(not_an_integer)`, there's another way of casting to integer that may or may not be more what you want. It's `not_an_integer.to_i`. You can read about [the subtle difference in behavior here](http://stackoverflow.com/questions/10093493/difference-between-integervalue-and-value-to-i#answer-10093533).
- I would probably have `Game.start` automatically assign `next_id` to the `id`, so that it's an internal implementation detail. Or are there times when you might want to start an existing game?

### More "railsy" routes

Instead of:

``` ruby
scope '/games' do
  post '/' => 'games#create'
  scope '/:gameId' do
    get  '/' => 'games#show', as: :games
    scope '/pegs' do
      scope '/:pegId' do
        put '/' => 'pegs#update'
      end
    end
  end
end
```

Which produces these routes:

```
Prefix Verb URI Pattern                          Controller#Action
       POST /games(.:format)                     games#create
 games GET  /games/:gameId(.:format)             games#show
       PUT  /games/:gameId/pegs/:pegId(.:format) pegs#update
```

Rails routes helpers actually allow you to do (almost) the same with just:

``` ruby
resources :games, only: [:show, :create] do
  resources :pegs, only: [:update]
end
```

That will produce:

```
  Prefix Verb  URI Pattern                        Controller#Action
game_peg PATCH /games/:game_id/pegs/:id(.:format) pegs#update
         PUT   /games/:game_id/pegs/:id(.:format) pegs#update
   games POST  /games(.:format)                   games#create
    game GET   /games/:id(.:format)               games#show
```

You may notice that `gameId` and `pegId` have been given different names, and the route helper prefixes are a little different, but you can update your other code accordingly.

**Comment by egillespie** [comment]
Thanks for the feedback! Here's some context that explains why I went against the grain of the Ruby way in some cases.

### General Ruby style stuff and random notes

> - For variable names, `snake_case` is prefered in Ruby rather than `camelCase`

I modeled my JSON before writing any code, and now I also have all of the client JavaScript expecting JSON to be structured that way so it made more sense in these ports to avoid changes to the client code.

> - In your models, `id` is usually preferred over something like `pegId`, as it's implied that the `id` is of the `Peg` when it's part of a `Peg` instance. 

This comes from ugly habits made from working on enterprise systems for too long. Hopefully I'll get more opportunities to design new APIs and steer away from that way of thinking soon. :)

> - Semicolons at end of lines not necessary - only when you want multiple statements on the same line (e.g. `a = 1; b = 2` or `def; some_method_name; end`)

D'oh. Fixed!

> - I would delete the unused files in `app/views`. I think only `static/home.html.erb` and `layouts/application.html.erb` are being used.

Removed.

> - The last line of a Ruby method is implicitly returned, so most people choose not to include a `return` there. This one is a little more subjective though.

Yeah, I'm pretty inconsistently taking advantage of that feature of the language. I tried to be consistent with a method, and didn't want to restructure some of those methods to only have one return point.

> - Apart from `Integer(not_an_integer)`, there's another way of casting to integer that may or may not be more what you want. It's `not_an_integer.to_i`. You can read about [the subtle difference in behavior here](http://stackoverflow.com/questions/10093493/difference-between-integervalue-and-value-to-i#answer-10093533).

I wanted to distinguish between numeric and non-numeric values in my validation, so I used `Integer(x)` so I could treat non-integers as `nil`. `x.to_i` looked promising until I saw that it returns 0 for non-integers.

> - I would probably have `Game.start` automatically assign `next_id` to the `id`, so that it's an internal implementation detail. Or are there times when you might want to start an existing game?

**TLDR** There's a reason, but it requires lots of context.

This is something I struggled with when porting Pegger to Ruby. I was thinking about storing games in a database eventually and writing a "Relational vs. NoSQL" follow-up article.

In some relational databases (namely, Oracle), it's not uncommon to use a sequence to generate unique IDs to be inserted into tables. This means there are two database operations that cause changes to the database: get the next sequence (which atomically increases the sequence), and the insert statement. I organized the database operations this way in the Java code.

This seemed like it would be a useful point to explain, so in the Java Pegger, I had a separate class for game operations so that I could combine these two database operations into a single database transaction. Express doesn't really have an opinion on where database operations go, so I was able to do something similar in the Node version of Pegger. Rails does seem to encourage database operations to go with the models, nor do I have a firm understanding of how transactions work in Rails, nor was it pertinent to my current effort to port Pegger to Rails, so that's why I left it this way. 

### More "railsy" routes

I like this way better and am totally changing what I've written!

**Comment by chrisvfritz** [comment]
Thanks for the greater context! Oracle... grrr. 😛
